File: robowflex_library/scripts/fetch_test.cpp
/* Author: Zachary Kingston */

#include <robowflex_library/builder.h>
#include <robowflex_library/detail/fetch.h>
#include <robowflex_library/scene.h>
#include <robowflex_library/trajectory.h>
#include <robowflex_library/util.h>

using namespace robowflex;

/* \file fetch_test.cpp
 * A simple script that demonstrates motion planning with the Fetch robot. The
 * resulting trajectory is output to a YAML file. This file can be visualized
 * using Blender. See the corresponding robowflex_visualization readme.
 */

static const std::string GROUP = "arm_with_torso";

int main(int argc, char **argv)
{
    // Startup ROS
    ROS ros(argc, argv);

    // Create the default Fetch robot.
    auto fetch = std::make_shared<FetchRobot>();
    fetch->initialize();

    // Create an empty scene.
    auto scene = std::make_shared<Scene>(fetch);

    // Create the default planner for the Fetch.
    auto planner = std::make_shared<OMPL::FetchOMPLPipelinePlanner>(fetch, "default");
    planner->initialize();

    // Sets the Fetch's base pose.
    fetch->setBasePose(1, 1, 0.5);

    // Sets the Fetch's head pose to look at a point.
    fetch->pointHead({2, 1, 1.5});

    // Create a motion planning request with a pose goal.
    MotionRequestBuilder request(planner, GROUP);
    fetch->setGroupState(GROUP, {0.05, 1.32, 1.40, -0.2, 1.72, 0.0, 1.66, 0.0});  // Stow
    request.setStartConfiguration(fetch->getScratchState());

    fetch->setGroupState(GROUP, {0.265, 0.501, 1.281, -2.272, 2.243, -2.774, 0.976, -2.007});  // Unfurl
    request.setGoalConfiguration(fetch->getScratchState());

    request.setConfig("RRTConnect");

    // Do motion planning!
    planning_interface::MotionPlanResponse res = planner->plan(scene, request.getRequest());
    if (res.error_code_.val != moveit_msgs::MoveItErrorCodes::SUCCESS)
        return 1;

    // Create a trajectory object for better manipulation.
    auto trajectory = std::make_shared<Trajectory>(res.trajectory_);

    // Output path to a file for visualization.
    trajectory->toYAMLFile("fetch_path.yml");

    return 0;
}

File: robowflex_library/fetch_test.py
from sys import argv

import robowflex_library as rf

GROUP = "arm_with_torso"

DEFAULT_ADAPTERS = [
    "default_planner_request_adapters/AddTimeParameterization",
    "default_planner_request_adapters/FixWorkspaceBounds",
    "default_planner_request_adapters/FixStartStateBounds",
    "default_planner_request_adapters/FixStartStateCollision",
    "default_planner_request_adapters/FixStartStatePathConstraints"
]

# Startup ROS
print('Starting ROS...')
ros = rf.ROS(len(argv), argv, "robowflex", 1)
print('ROS running!')

# Create the default Fetch robot
print('Creating Fetch robot...')
fetch = rf.FetchRobot()
fetch.initialize(True)

# Create an empty scene
print('Creating scene...')
scene = rf.Scene(fetch)

# Create the default planner for the Fetch
print('Creating planner...')
planner = rf.FetchOMPLPipelinePlanner(fetch, "default")
planner.initialize(rf.Settings(), DEFAULT_ADAPTERS)

# Set the Fetch's base pose
print('Posing Fetch...')
fetch.set_base_pose(1, 1, 0.5)

# Set the Fetch's head pose to look at a point
fetch.point_head([2, 1, 1.5])

# Create a motion planning request with a pose goal
print('Creating planning request...')
request = rf.MotionRequestBuilder(planner, GROUP, '')
fetch.set_group_state(GROUP, [0.05, 1.32, 1.40, -0.2, 1.72, 0.0, 1.66, 0.0])
request.set_start_configuration(fetch.get_scratch_state())

# Unfurl
fetch.set_group_state(GROUP, [0.265, 0.501, 1.281, -2.272, 2.243, -2.774, 0.976, -2.007])
request.set_goal_configuration(fetch.get_scratch_state())
request.set_config("RRTConnect")

# Do motion planning!
print('Motion planning...')
result = planner.plan(scene, request.get_request())
# NOTE: We don't export the MoveIt messages library; could probably get the below constant from the
# Python bindings but I'm lazy
if result.error_code_.val != 1:
  raise RuntimeError(f'Motion planning failed with code: {result.error_code_.val}')

# Create a trajectory object for better manipulation
print('Exporting trajectory...')
trajectory = rf.Trajectory(result.trajectory_)

# Output path to a file for visualization
trajectory.to_yaml_file("fetch_path.yml")
